lib/streamlit/state/widgets.py

```diff
@@ -67,29 +72,44 @@ class NoValue:
     pass
 
 
-@attrs(auto_attribs=True)
+@attr.s(auto_attribs=True)
 class Widget:
-    state: WidgetState
-
-    def type(self) -> str:
+    id: str
+    state: Optional[WidgetState] = None
+    callback: Optional[WidgetCallback] = None
+    deserializer: WidgetDeserializer = lambda x: x
+    callback_args: Optional[WidgetArgs] = None
+    callback_kwargs: Optional[WidgetKwargs] = None
+
+    @property
+    def type(self) -> Optional[str]:
+        if self.state is None:
+            return None
         return self.state.WhichOneof("value")
 
-    def id(self):
-        return self.state.id
-
+    @property
     def value(self) -> Any:
-        if self.type() == "json_value":
-            return json.loads(getattr(self.state, self.type()))
+        if self.type is None:
+            return self.deserializer(None)
```

Comment:
Is there a reason for the deserializers to have to deal with None? If they all would just return None then skipping that part makes them easier to deal with, but if they return meaningful values for None then the current way is right.

Reply:
Calling `deserializer` on `None` ensures that the correct default is returned when `WidgetManager.get_widget_value()` on a widget that hasn't had a `state` attr set yet (which may happen when a widget is first registered on a script run and hasn't been changed on the frontend yet)